packages/node/src/integrations/localvariables/localvariables-async.ts

```diff
@@ -0,0 +1,287 @@
+import type { Event, EventProcessor, Exception, Hub, Integration, StackFrame, StackParser } from '@sentry/types';
+import { LRUMap, logger } from '@sentry/utils';
+import type { Debugger, InspectorNotification, Runtime } from 'inspector';
+import type { NodeClient } from '../../client';
+
+import type { NodeClientOptions } from '../../types';
+import type { FrameVariables, Options, PausedExceptionEvent, RateLimitIncrement, Variables } from './common';
+import { createRateLimiter, functionNamesMatch, hashFrames, hashFromStack } from './common';
+
+type Session = {
+  connect: () => void;
+  new (): Session;
+
+  post(method: 'Debugger.pause' | 'Debugger.resume' | 'Debugger.enable' | 'Debugger.disable'): Promise<void>;
+  post(method: 'Debugger.setPauseOnExceptions', params: Debugger.SetPauseOnExceptionsParameterType): Promise<void>;
+  post(
+    method: 'Runtime.getProperties',
+    params: Runtime.GetPropertiesParameterType,
+  ): Promise<Runtime.GetPropertiesReturnType>;
+
+  on(
+    event: 'Debugger.paused',
+    listener: (message: InspectorNotification<Debugger.PausedEventDataType>) => void,
+  ): Session;
+
+  on(event: 'Debugger.resumed', listener: () => void): Session;
+};
+
+async function unrollArray(session: Session, objectId: string, name: string): Promise<Variables> {
+  const properties = await session.post('Runtime.getProperties', { objectId, ownProperties: true });
+
+  return {
+    [name]: properties.result
+      .filter(v => v.name !== 'length' && !isNaN(parseInt(v.name, 10)))
+      .sort((a, b) => parseInt(a.name, 10) - parseInt(b.name, 10))
+      .map(v => v?.value?.value),
+  };
+}
+
+async function unrollObject(session: Session, objectId: string, name: string): Promise<Variables> {
+  const properties = await session.post('Runtime.getProperties', { objectId, ownProperties: true });
+
+  return {
+    [name]: properties.result
+      .map<[string, unknown]>(v => [v.name, v?.value?.value])
+      .reduce((obj, [key, val]) => {
+        obj[key] = val;
+        return obj;
+      }, {} as Variables),
+  };
+}
+
+function unrollOther(prop: Runtime.PropertyDescriptor): Variables {
+  if (prop?.value?.value) {
+    return {
+      [prop.name]: prop.value.value,
+    };
+  } else if (prop?.value?.description && prop?.value?.type !== 'function') {
+    return {
+      [prop.name]: `<${prop.value.description}>`,
+    };
+  }
+
+  return {};
+}
+
+async function getLocalVariables(session: Session, objectId: string): Promise<Variables> {
+  const properties = await session.post('Runtime.getProperties', { objectId, ownProperties: true });
+  let variables = {};
+
+  for (const prop of properties.result) {
+    if (prop?.value?.objectId && prop?.value.className === 'Array') {
+      const id = prop.value.objectId;
+      variables = { ...variables, ...(await unrollArray(session, id, prop.name)) };
```

Comment:
This copies `...variables` every time, and will cause a rise in memory usage. We should avoid making unnecessary copies.

I believe this might work:

```suggestion
      Object.assign(variables, await unrollArray(session, id, prop.name));
```

Reply:
I'll just pass `variables` is as a reference and the functions can add properties to it!